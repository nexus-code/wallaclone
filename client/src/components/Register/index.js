import Register from './Register';
import withSession from '../../hocs/withSession'

export default withSession(Register);