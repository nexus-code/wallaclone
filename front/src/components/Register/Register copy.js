import React, { useReducer } from "react";

import Canvas from '../Canvas/Canvas';

import { Form, Button } from 'react-bootstrap';
import { toast } from 'react-toastify';

/**
 *  session user handler
 */

export default function Register({ user, setUser, logout }) { 

    // uses toast to ui add notifications   
    const notifyWarning = (warning) => toast.warning(warning);
    ///

    const status = typeof (user) === 'undefined' ? 0 : 1; // 0 'REGISTER' : 1 'EDIT';
    const pageTitle = status ? 'Edit profile' : 'Register user';


    const [userInput, setUserInput] = useReducer(
        (state, newState) => ({ ...state, ...newState }),
        {
            name: status ? user.name : '',
            surname: status ? user.surname : '',
        }
    );

    const handleLogout = () => {
        setUserInput({name: '', surname: ''});
        logout();
    }
    const logoutButton = status ? <Button variant="secondary" className="float-right logoutButton" onClick={handleLogout}>Logout</Button> : '';
    

    const handleChange = event => {

        const name = event.target.name;
        const newValue = event.target.value;
        setUserInput({ [name]: newValue });
    }

    const handleSubmit = event => {
        event.preventDefault();

        if (userInput.name.trim().length <= 3) {
            notifyWarning('The name must be bigger than 3 characters');
            return;
        }

        if (userInput.surname.trim().length <= 3) {
            notifyWarning('The surrname must be bigger than 3 characters');
            return;
        }

        setUser({ name: userInput.name, surname: userInput.surname});
    }
        
    return (
        <Canvas>
            <div style={{ padding: "20px", maxWidth: "420px", margin: "50px auto" }}>
                <h2>{ pageTitle }</h2>
                <Form onSubmit = { handleSubmit }>
                    <Form.Group controlId="formGroupname" >
                        <Form.Label>Name</Form.Label>
                        <Form.Control name="name" placeholder="Enter name" value={ userInput.name } onChange={ handleChange } />
                    </Form.Group>
                    <Form.Group controlId="formGroupsurname" >
                        <Form.Label>Surname</Form.Label>
                        <Form.Control name="surname" placeholder="Enter surname" value={ userInput.surname } onChange={ handleChange } />
                    </Form.Group>


                    <Button variant="primary" className="submitButton" type="submit">
                        Save
                    </Button>
                    
                    { logoutButton }

                </Form>
            </div>
        </Canvas>
    );
}