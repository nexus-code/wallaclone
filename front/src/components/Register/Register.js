import React from "react";
import Canvas from '../Canvas/Canvas';
import { useForm } from "react-hook-form";
import i18n from '../../i18n';
import { useTranslation } from 'react-i18next';

export default function Register({ user, setUser, logout }) { 

    const { t } = useTranslation();

    const status = typeof (user) === 'undefined' ? 0 : 1; // 0 'REGISTER' : 1 'EDIT';
    const pageTitle = status ? 'Edit profile' : 'Register user';

    const { register, handleSubmit, reset, errors } = useForm({ defaultValues: user });
    const onSubmit = data => {

        changeLanguage(data.language)
        setUser(data);
    };

    const changeLanguage = (lng) => {
        i18n.changeLanguage(lng);
    }

    const handleLogout = () => {
        
        reset();    // does not empty entries !!
        logout();
    }
    const logoutButton = status ? <button variant="secondary" className="float-right logoutbutton" onClick={handleLogout}>{t('Logout')}</button> : '';
    
    function validator (field, minLength, maxLength) {
        return {
            required: t(field) +` ${t('is required')}`,
            minLength: {
                value: minLength,
                message: `Min length is ${minLength}`
            },
            maxLength: {
                value: maxLength,
                message: `Max length is ${maxLength}`
            }
        }
    };
      
    return (
        <Canvas>
            <div style={{ padding: "20px", maxWidth: "420px", margin: "50px auto" }}>
                <h2>{ t(pageTitle) }</h2>
                <form onSubmit={handleSubmit(onSubmit)}>
                    <label>{ t('Username') }</label>
                    <input 
                        name="username" 
                        placeholder="{{ t(Insert your unique username) }}" 
                        ref={register(validator('username', 3, 25))}  
                    />
                    {errors.username && <p>{errors.username.message}</p>}
                    <label>{ t('Password') }</label>
                    <input
                        type="password"
                        name="password"
                        placeholder="Insert your password"
                        ref={register(validator('password', 3, 25))}  
                    />
                    {errors.password && <p>{errors.password.message}</p>}
                    <label>{ t('Email') }</label>
                    <input
                        type="text"
                        name="email"
                        ref={register({
                            required: "Email is required",
                            pattern: {
                                value: /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
                                message: t("Invalid email address")
                            }
                        })}
                    />
                    {errors.email && <p>{errors.email.message}</p>}

                    <label>{ t('Language') }</label>
                    <select ref={register} name="language">
                        <option value="en">English</option>
                        <option value="es">Castellano</option>
                    </select>

                    <label>{ t('Remember me?') }<input name="remember" type="checkbox" ref={register} /></label>
                    
                    <input type="submit" />
                </form>
                    
                { logoutButton }
                <br />
                <hr />
                <br />

            </div>
        </Canvas>
    );
}