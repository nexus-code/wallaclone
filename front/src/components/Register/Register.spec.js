import React from 'react';
import { shallow } from 'enzyme';

import Register from './Register';
// import { setUser } from '../../store/user/actions';

describe('Register', () => {
    
    const props = {
        user: {
            name: 'test name',
            surname: 'test surname',
        },
        setUser: jest.fn(),
        logout: jest.fn(),
    };
    
    const wrapper = shallow(<Register {...props} />);


    it('should render a Register component', () => {
        
        expect(wrapper).toMatchSnapshot();
    });

    describe('should do logout', () => {

        wrapper
            .find('.logoutButton')
            .simulate('click');

        it('should logout user', () => {
            expect(props.logout).toHaveBeenCalled();
        });
    });

    describe('should register user', () => {
        wrapper
        .find('[name="name"]')
        .simulate("change", { target: { value: props.user.name, name: "name" } });
        
        wrapper
            .find('[name="surname"]')
            .simulate("change", { target: { value: props.user.surname, name: "surname" } });
            
            
        
        const preventDefault = jest.fn();
        wrapper.find('Form').simulate('submit', { preventDefault });
        
        it('should prevent default Form submission', () => {
            
            expect(preventDefault).toHaveBeenCalled();
        });
        

        it('should save user data', () => {

                
            // console.log('wrapper.debug(FORM)', wrapper.find('Form').debug());

            expect(props.setUser).toHaveBeenCalledWith(
            
                expect.objectContaining(props.user)
            );

        });

    });
});
