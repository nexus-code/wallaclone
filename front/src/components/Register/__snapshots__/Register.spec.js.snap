// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should render a Register component 1`] = `
<Layout>
  <div
    style={
      Object {
        "margin": "50px auto",
        "maxWidth": "420px",
        "padding": "20px",
      }
    }
  >
    <h2>
      Edit profile
    </h2>
    <Form
      inline={false}
      onSubmit={[Function]}
    >
      <FormGroup
        controlId="formGroupname"
      >
        <FormLabel
          column={false}
          srOnly={false}
        >
          Name
        </FormLabel>
        <FormControl
          name="name"
          onChange={[Function]}
          placeholder="Enter name"
          value="test name"
        />
      </FormGroup>
      <FormGroup
        controlId="formGroupsurname"
      >
        <FormLabel
          column={false}
          srOnly={false}
        >
          Surname
        </FormLabel>
        <FormControl
          name="surname"
          onChange={[Function]}
          placeholder="Enter surname"
          value="test surname"
        />
      </FormGroup>
      <Button
        active={false}
        className="submitButton"
        disabled={false}
        type="submit"
        variant="primary"
      >
        Save
      </Button>
      <Button
        active={false}
        className="float-right logoutButton"
        disabled={false}
        onClick={[Function]}
        type="button"
        variant="secondary"
      >
        Logout
      </Button>
    </Form>
  </div>
</Layout>
`;
